packages/starknet-snap/src/rpcs/executeTxn.ts

```diff
@@ -0,0 +1,248 @@
+import type { Transaction } from '../types/snapState';
+import { VoyagerTransactionType } from '../types/snapState';
+import {
+  AddressStruct,
+  BaseRequestStruct,
+  AccountRpcController,
+  AuthorizableStruct,
+  confirmDialog,
+  InvocationStruct,
+  UniversalDetailsStruct,
+} from '../utils';
+import {
+  ACCOUNT_CLASS_HASH,
+  CAIRO_VERSION,
+  TRANSACTION_VERSION,
+} from '../utils/constants';
+import { addDialogTxt, getTxnSnapTxt } from '../utils/snapUtils';
+import {
+  createAccount,
+  executeTxn as executeTxnUtil,
+  getAccContractAddressAndCallData,
+  getEstimatedFees,
+  isAccountDeployed,
+} from '../utils/starknetUtils';
+
+export const ExecuteTxnRequestStruct = refine(
+  assign(
+    object({
+      address: AddressStruct,
+      invocations: array(InvocationStruct),
+      details: UniversalDetailsStruct,
+      abis: optional(any()),
+    }),
+    AuthorizableStruct,
+    BaseRequestStruct,
+  ),
+  'ExecuteTxnRequestStruct',
+  (value) => {
+    if (value.invocations.length === 0) {
+      return 'Invocations cannot be empty';
+    }
+    for (const invocation of value.invocations) {
+      if (invocation.type !== TransactionType.INVOKE) {
+        return `Invocations should be of type ${TransactionType.INVOKE} received ${invocation.type}`;
+      }
+    }
+    return true;
+  },
+);
+
+export const ExecuteTxnResponseStruct = object({
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  transaction_hash: string(),
+});
+
+export type ExecuteTxnParams = Infer<typeof ExecuteTxnRequestStruct> & Json;
+
+export type ExecuteTxnResponse = Infer<typeof ExecuteTxnResponseStruct>;
+
+const getInvokeCalls = (invocations: Invocations): Call[] => {
+  return invocations
+    .filter((invocation) => invocation.type === TransactionType.INVOKE)
+    .map((invocation: any) => invocation.payload as Call);
+};
+
+/**
+ * The RPC handler to execute a transaction.
+ */
+export class ExecuteTxnRpc extends AccountRpcController<
+  ExecuteTxnParams,
+  ExecuteTxnResponse
+> {
+  protected requestStruct = ExecuteTxnRequestStruct;
+
+  protected responseStruct = ExecuteTxnResponseStruct;
+
+  /**
+   * Execute the transaction request handler.
+   *
+   * @param params - The parameters of the request.
+   * @param params.address - The address of the signer.
+   * @param params.invocations - The invocations to execute (only invocations of type TransactionType.INVOKE)
+   * @param params.abis - The abis associated to invocations.
+   * @param params.details - The detail associated to the call.
+   * @returns The InvokeFunctionResponse as an `ExecuteTxnResponse`.
+   */
+  async execute(params: ExecuteTxnParams): Promise<ExecuteTxnResponse> {
+    return super.execute(params);
+  }
+
+  protected async handleRequest(
+    params: ExecuteTxnParams,
+  ): Promise<ExecuteTxnResponse> {
+    const { address, invocations, abis, details, enableAuthorize } = params;
+
+    const calls = getInvokeCalls(invocations as Invocations);
```

Comment:
why we need this method?

Reply:
- `getEstimatedFees` accepts `Invocations` and `UniversalDetails`
- `execute` accepts `calls or calls[]` and `UniversalDetails`

This method is to extract the invoke calls from the Invocations to get them in right format for the execute method.

Reply:
i see, but i think we dont need to use invocation for now, as it should be calls[] | calls, we can create a struct named OptionalArrayStruct, i will do it for you, you can leave it, but we can remove getInvokeCalls to align the exectueTxn as starknet.js?

Reply:
we need invocation, its for the estimation of fees. The starknet estimate fee bulk method we use requires type Invocation not type calls[] | call 

```ts
  ).estimateFeeBulk(txnInvocation, {
    ...invocationsDetails,
    skipValidate: false,
    blockIdentifier: BlockIdentifierEnum.Latest,
  });
  ```

Reply:
We can do the opposite of getInvokeCalls, e.g. accepts calls[] | calls and build the invocation object with that.

Reply:
i think that on executeTxn , we better only accept calls[] | calls, as it is use by get starknet, if the type is not match, it is hard for the dev

Reply:
Agree, refactored to accept calls[] | calls like starknet